cranelift/codegen/src/isa/unwind.rs

```diff
@@ -86,3 +89,89 @@ pub mod input {
         pub initial_sp_offset: u8,
     }
 }
+
+/// Unwind pseudoinstruction used in VCode backends: represents that
+/// at the present location, an action has just been taken.
+///
+/// VCode backends always emit unwind info that is relative to a frame
+/// pointer, because we are planning to allow for dynamic frame allocation,
+/// and because it makes the design quite a lot simpler in general: we don't
+/// have to be precise about SP adjustments throughout the body of the function.
+///
+/// Unwind pseudoinstructions are emitted inline by ABI code as it generates
+/// a prologue.
+///
+/// Thus, for the usual case, a prologue might look like (using x64 as
+/// an example):
+///
+/// ```plain
+/// push rbp
+/// unwind PushOldFP sp_offset_to_entry=8
+/// mov rbp, rsp
+/// unwind CreateFPFrame (BP=sp; [BP] = oldBP)
+/// sub rsp, 32
+/// mov [rsp+0], r12
+/// unwind SaveReg [BP-32], r12
+/// mov [rsp+8], r13
+/// unwind SaveReg [BP-24], r13
+/// ...
+/// ...
+/// mov rsp, rbp
+/// pop rbp
+/// unwind PopOldFP sp_offset_to_entry=0
+/// ret
+/// unwind Ret
+/// ```
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+pub enum UnwindInst {
+    /// The frame-pointer register for this architecture has just been
+    /// pushed to the stack. The FP has not been set to this frame yet.
+    /// The push decremented SP so that it is now `sp_offset_to_caller`
+    /// bytes below SP at the caller's callsite.
+    PushOldFP {
+        /// The offset from the current SP (after push) to the SP at
+        /// caller's callsite.
+        sp_offset_to_caller: u8,
+    },
+    /// The frame-pointer register for this architecture has just been popped
+    /// from the stack. The frame is now defined by SP again, and as above for
+    /// `PushOldFP`, `sp_offset_to_caller` gives the distance from current SP
+    /// (after pop) to SP at the caller's callsite.
+    PopOldFP {
+        /// The offset from the current SP (after pop) to the SP at
+        /// caller's callsite.
+        sp_offset_to_caller: u8,
+    },
+    /// The frame-pointer register for this architecture has just been
+    /// set to the current stack location, and the word at FP is the
+    /// previous FP value.
+    ///
+    /// The base of the frame is at `FP - fp_offset`. Note that only
+    /// positive offsets from the base of the frame are allowed for
+    /// `SaveReg`, but the FP register itself can point to any part
+    /// of the frame.
+    CreateFPFrame {
```

Comment:
Ok so after reading the various ABI implementations I think I understand a little better, but not entirely. It seems like this is primarily around for Windows and calculating adjustments on Unix, right?

For the Unix case (which I think I understand better) this seems like it should be something like `DeclareCobberFrameSize { size: u8 }` where it's basically just declaring that all saves to come later are relative offsets within a frame of size `size`, allocated at the top of the stack just after the frame pointer was pushed.

For Windows I'm much more unfamiliar with the unwinding stuff. I'm a bit confused, though, on why everything is `u8` here. It seems to be because of Windows but the [exception codes](https://docs.microsoft.com/en-us/cpp/build/exception-handling-x64?view=msvc-160) all have short/far variants which seems like it's trying to account for offsets which are larger than `u8` perhaps, but I may be misunderstanding something. For Windows though it seems like this should be two instructions, where one signifies the `mov %rsp, %rbp` happened and the other happens after the stack decrement.

I'll admit though I'm pretty confused on Windows, and I probably just need to sit down and read through their documentation some more. I'm not sure if we're doing RSP-relative addressing or whatever the second mode is mentioned [at the top of this](https://docs.microsoft.com/en-us/cpp/build/exception-handling-x64?view=msvc-160#unwind-operation-code) where something is FP relative. I always seem to get lost in the details thinking about stuff like this.

Reply:
So WIndows forces our hand in a few ways here:

* The way that clobbers are located is as a *positive* offset from the "frame", wherever we choose to define the frame. Since we place `RBP`/`FP` just below the return address so that we have known offset addressing to stack args, and we place clobbers below `RBP`, we have to either define the frame in terms of `RSP` or as some offset below `RBP`. The former is difficult for several reasons but basically our nominal-SP tracking would not interact well with unwind info due to the order we compute and finalize things. So we define the frame in terms of `RBP`, but `RBP` is at the *top* of it.
* This is what `fp_offset` means (it comes directly from the "frame register offset" field in `UNWIND_INFO`, which is also the source of the `u8`-sized limit): when we declare the frame, its base is at `RBP - fp_offset`. Equivalently, this is saying that the frame pointer is at the "frame pointer offset" into the frame.
* Note that on SysV platforms, the frame is defined by the "CFA" (call-frame address) which is spec-defined to be the SP value at entry, so instead `fp_offset` is used to adjust the offsets for clobber-saves.
* The `fp_offset` size limit (actually 240 bytes) is the reason that clobbers had to be moved to just below `RBP` as well; otherwise we could not define a frame far enough away from `RBP` to straddle our other fixed-frame slots.

Reply:
Regarding `fp_offset` in particular, I'll go over the doc comments some more and see if I can add detail. Perhaps what we really need is just a small tutorial on unwind formats on Windows and SysV and how frames are computed in both cases :-)

Reply:
Ok thanks for the info! This helps dispel some of my confusion. I'm familiar with the CFA in dwarf but the "cfa equivalent" was actually in the other direction on Windows and I was getting confused how these two things map to each other. I think for now some extra documentation for how the field is expected to be interpreted on Windows and Unix should be fine myself.

But for the `u8`-ness, to make sure I understand, we're declaring to Windows "all register saves, which occur at positive offsets, are relative to XX" where XX is `RBP-thing` and `thing` is limited to 8 bits? We then have an *additional* requirement later on that when we save a register it can be saved at most 240 bytes away? (I couldn't actually find where that second restriction arose myself)

Reply:
Re: offset restrictions, only the former (`RBP - offset`); once we establish the frame we can declare clobber-saves at any 32-bit offset. But combining this with the fact that we place `RBP` *above* clobber-saves gives us the limit on clobber offsets.

In hindsight, perhaps it's better to have explicit asserts that all offsets are in-range for the unwind info we're generating, and give these offset fields a `u32` or so; that would document these limits more explicitly I think!

Reply:
(Err, "any positive 32-bit offset" above.)

Reply:
Alternatively we could stop establishing a frame pointer until such a time we actually need one on Windows (the save offsets encoded in the unwind information would then be positive offsets from RSP rather than FP-[0..240], thereby eliminating the "saves have to be near enough to the frame pointer establishment" problem).  That's a whole can of worms, though, I imagine.

Reply:
In fact, the old unwind info emitted for Windows explicitly didn't encode establishing a frame pointer (despite the prologue doing so) to side step the 240 offset limitation, thereby encoding all of the register saves as positive offsets from the stack pointer.  This would obviously break when the frame pointer is required, such as dynamically-sized frames.

Reply:
Aha ok that makes a lot of sense now, where the 240 comes from. I do think there's definitely room for improvement here, but there's no need to happen in this PR I think. I'd be happy with some more comments myself, and then seems reasonable to open tracking issues if necessary.

Reply:
So I've revamped the doc-comment for this enum and added the "offset upward / offset downward" dual fields for the CreateNewFrame pseudo-inst. Hopefully this is a bit more clear now!

Agreed that we could eventually look into `-fomit-frame-pointer` codegen, though that's a deeper rabbithole in the ABI code (as is `alloca` support); one bridge at a time :-)